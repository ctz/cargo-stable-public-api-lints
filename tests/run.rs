use std::mem;
use std::process::{Command, Output};

#[test]
fn non_exhaustive_enum_fixed() {
    insta::assert_debug_snapshot!(run(
        "tests/non_exhaustive/",
        "exhaustive-enum",
        "enum_fixed"
    ));
}

#[test]
fn non_exhaustive_enum_warn() {
    insta::assert_debug_snapshot!(run("tests/non_exhaustive/", "exhaustive-enum", "enum_warn"));
}

#[test]
fn non_exhaustive_enum_allow() {
    insta::assert_debug_snapshot!(run(
        "tests/non_exhaustive/",
        "exhaustive-enum",
        "enum_allow"
    ));
}

#[test]
fn non_exhaustive_struct_fixed() {
    insta::assert_debug_snapshot!(run(
        "tests/non_exhaustive/",
        "exhaustive-struct",
        "struct_fixed"
    ));
}

#[test]
fn non_exhaustive_struct_warn() {
    insta::assert_debug_snapshot!(run(
        "tests/non_exhaustive/",
        "exhaustive-struct",
        "struct_warn"
    ));
}

#[test]
fn non_exhaustive_struct_allow() {
    insta::assert_debug_snapshot!(run(
        "tests/non_exhaustive/",
        "exhaustive-struct",
        "struct_allow"
    ));
}

#[test]
fn drop_struct_warn() {
    insta::assert_debug_snapshot!(run("tests/drop/", "drop", "struct_warn"));
}

#[test]
fn sorted_enum() {
    insta::assert_debug_snapshot!(run("tests/sorted_enum/", "sorted-enum", "default"));
}

fn run(path: &str, lint_id: &str, feature: &str) -> Output {
    let mut output = Command::new("cargo")
        .arg("-q")
        .arg("run")
        .arg("--")
        .arg("--features")
        .arg(feature)
        .arg("--allow")
        .arg("all")
        .arg("--warn")
        .arg(lint_id)
        .arg(format!("{path}/Cargo.toml"))
        .output()
        .unwrap();

    // currently the rustdoc subprocess writes unstable output
    // to stderr
    println!(
        "stderr: {}",
        String::from_utf8(mem::take(&mut output.stderr)).expect("stderr is not valid utf8")
    );
    println!(
        "stdout:\n{}",
        String::from_utf8(output.stdout.clone()).expect("stdout is not valid utf8")
    );
    output
}
