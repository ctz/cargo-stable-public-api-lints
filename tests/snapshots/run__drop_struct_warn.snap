---
source: tests/run.rs
expression: "run(\"tests/drop/\", \"drop\", \"struct_warn\")"
---
Output {
    status: ExitStatus(
        unix_wait_status(
            256,
        ),
    ),
    stdout: "------------------------------------------------------------\nIt is a breaking change for a type to become `Drop` later, and this also happens\n\"magically\" when any type reachable through any field or variant becomes `Drop`.\nThis includes the types of private fields!\n\nTherefore it is prudent to `impl Drop` up-front to allow the maximum amount of\nfuture non-breaking changes, both in your code and the code of your dependencies.\n\nAnnotate this type with `#[cfg_attr(spal, allow(spal::drop))]` to ignore this lint\n------------------------------------------------------------\n\n⚠\u{fe0f} enum `pub enum drop::CopyEnum` should impl Drop\n  at src/lib.rs:21\n\n⚠\u{fe0f} enum `pub enum drop::NoDropEnum` should impl Drop\n  at src/lib.rs:10\n\n⚠\u{fe0f} struct `pub struct drop::CopyStruct` should impl Drop\n  at src/lib.rs:16\n\n⚠\u{fe0f} struct `pub struct drop::DropGlueStruct` should impl Drop\n  at src/lib.rs:1\n\n⚠\u{fe0f} struct `pub struct drop::DropPublic` should impl Drop\n  at src/lib.rs:34\n\n⚠\u{fe0f} struct `pub struct drop::NoDropStruct` should impl Drop\n  at src/lib.rs:6\n\n",
    stderr: "",
}
